src-built-in/components/UIAPI/reducers/toolbar.ts

```diff
@@ -0,0 +1,27 @@
+// import { loop, Cmd } from 'redux-loop';
+
+// import { ActionTypes } from '../types';
+// import { loadPinnedItems } from '../effects/toolbar';
+
+// const initialState = {
```

Comment:
I think the state design should be reverse the hierarchy here. Eventually we want to be able to handle *any* type of favorite. So it should be a list of favorites. A favorite itself should have a type which can be {"workspace"|"component"}. This will allow us to add more types in the future.

Reply:
Yes when I was drawing the diagram I realized it should be something like this: 
![state](https://user-images.githubusercontent.com/50024509/73987006-13929f00-490d-11ea-8e83-60295cea976c.jpg)

Reply:
Not sure why it's rotated to the left. If you click on the image it will open another page with the right orientation

Reply:
You've already taken a step towards a normalized schema, which is great. Probably you're going to need unique identifiers.

```
items : {
   "GSD1212SSD" : {
   "label": "Welcome component",
   "action": "SWITCH_WORKSPACE",
   "payload" : {}
   }
},
order : ["GSD1212SSD","XSD12323A"
]

@d4hines was investigating normalization best practices for redux:
https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape/

Reply:
You're right, we'll need IDs for every favorite component. Normalized state should be very useful. I'm aligning my current design to that practice.